src/main.rs

```diff
@@ -149,7 +193,25 @@ impl State {
     ///
     /// No change is made to `self` if an error occurs.
     fn do_move(&mut self, mov: Move) -> Result<NextStep, HanoiError> {
-        unimplemented!()
+        //this_disk will come in useful later, so we bind it
+		let this_disk = self.peek_disk(mov.from);
+		//if no disk, throw an error - can't try! here because peek_disk returns an option
+		if this_disk.is_none() {
+			Err(HanoiError::EmptyFrom(mov.from))
+		}
+		else {
+			//if the try fails here, return the error early
+			try!(self.push_disk(mov.to,this_disk.unwrap()));
+			//if the try worked, we can kill the disk we're supposed to have moved
+			self.pop_disk(mov.from);
+			//check for a win
+			if self.done() {
+				Ok(NextStep::Win)
+			}
+			else {
+				Ok(NextStep::Continue)
+			}
+		}
```

Comment:
It would be more idomatic to use `match` rather than `if/else`with an unwrap. Generally avoiding unwrap is good, because it makes it easier to very that code will execute without panicking.